### PR TITLE
PLANNER-1821 & PLANNER-1873: Add Constraint Provider for Coach Shuttle Gathering

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
### JIRA

https://issues.redhat.com/browse/PLANNER-1873
https://issues.redhat.com/browse/PLANNER-1821

### Checklist
- [ ] Documentation updated if applicable.
- [ ] Upgrade recipe provided if applicable.

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add comment: <b>Jenkins run fdb</b>
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

Replaces https://github.com/kiegroup/optaplanner/pull/691 and https://github.com/kiegroup/optaplanner/pull/793

Coach Capacity was a fun one; it would be easy if we had an accumlate that works on empty streams, but since we don't, I need to implement it as three constraints: One when no streams are empty, one that handles when coach is overfilled without shuttle, and one to remove double counting.

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/CoachPassengerCountTotalUpdatingVariableListener.java
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/CoachPassengerCountTotalUpdatingVariableListener.java:41]: Please follow the naming of the other VariableListeners for consistency across examples, such as ArrivalTimeUpdatingVariableListener. These 2 methods would be "update..." and be below the before/after calls.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/CoachPassengerCountTotalUpdatingVariableListener.java
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/CoachPassengerCountTotalUpdatingVariableListener.java:55]: method naming: increase/decrease what?
Also, I wouldn't extract them into separate methods, just call true/false directly from before/after.

**Review by ge0ffrey — APPROVED** [review]
Looks good. See some naming/method order suggestions, see comments.

Before merging, please validate that this is a pure refactor: check if the current master (DRL without this new shadow var) and this branch (CS with this new shadow var) deliver the same line in DEBUG logging on Local Search step 100 (except for the time spent of course). Chris, please try that out and report your findings here. If they don't match, let's figure out why.

**Review by triceo — CHANGES_REQUESTED** [review]
After Geoffrey's comments are addressed and the solution is proven in the way that he suggests, changes need to be made to the solver config to restore it to its original form.

**Comment by Christopher-Chianelli** [comment]
@ge0ffrey the 100 LS line was identical. Performance.... kinda bad
```
Master using old Drools
12:52:26.696 [l-3-thread-1] DEBUG     LS step (100), time spent (498), score (0hard/-400050soft),     best score (0hard/-384040soft), accepted/selected move count (1/1), picked move (S5 {B2 -> B1}).

New Branch using new Drools w/ Shadow Variables
13:14:19.973 [l-3-thread-1] DEBUG     LS step (100), time spent (2384), score (0hard/-400050soft),     best score (0hard/-384040soft), accepted/selected move count (1/1), picked move (S5 {B2 -> B1}).

New Branch using CP + Shadow Variables
13:08:35.521 [l-3-thread-1] DEBUG     LS step (100), time spent (3169), score (0hard/-400050soft),     best score (0hard/-384040soft), accepted/selected move count (1/1), picked move (S5 {B2 -> B1}).
```

I tested the new Drools and ConstraintProvider against the EasyScoreCalculator on full assert; no issues came up.
2384 / 498 = 4.787..., so almost 5 time slower using new Drools
3169 / 498 = 6.363..., so over 6 times slower using ConstraintProvider.

My guess: the variable update listener eats a lot of time and is called a lot (although it only increments/decrements a value)

**Comment by triceo** [comment]
Actually, 30 % down for CS-D ain't that bad, considering that this is before the new groupBy implementation.

Did you see this results with `FULL_ASSERT` enabled? I'd argue that this would not really be relevant, as constraint matches add a lot of that wouldn't otherwise be used. (Although I could also argue that this would make relative results better, not worse.)

**Comment by Christopher-Chianelli** [comment]
@triceo result is with FULL_ASSERT enable.... probably should rerun

**Comment by Christopher-Chianelli** [comment]
New drools w/o full assert
```
14:10:33.371 [l-3-thread-1] DEBUG     LS step (100), time spent (495), score (0hard/-400050soft),     best score (0hard/-384040soft), accepted/selected move count (1/1), picked move (S5 {B2 -> B1}).
```
About same speed (compared to old drools)

Constraint Provider w/o full assert
```
14:12:23.113 [l-3-thread-1] DEBUG     LS step (100), time spent (711), score (0hard/-400050soft),     best score (0hard/-384040soft), accepted/selected move count (1/1), picked move (S5 {B2 -> B1}).
```
So about 43% performance loss (compared to old drools)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/Shuttle.java
Comment (Christopher-Chianelli) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/Shuttle.java:60]: Can't do - leads to double counting it seems (despite code being exactly the same); was able to put it in an abstract class and have two variable listeners extend it (the abstract method is literally an instanceof check)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/BusPassengerCountTotalUpdatingVariableListener.java
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/BusPassengerCountTotalUpdatingVariableListener.java:74]: We can probably improve this:

    int quantity = busStop.getPassengerQuantity();
    if (quantity == 0) {
        return;
    }
    int difference = increase ? quantity : -quantity;

If this code is called often, this will bring a small perf improvement.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java:40]: Let's `import static ...ConstraintCollectors...`
AFAIK the other constraint providers do it too.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java:42]: Not sure what the convention here is, but maybe `import static ... HardSoftLongScore...` as well?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java:74]:
```suggestion
    /* 
       Correct the double counting
       Explanation: groupBy is like accumulate, but it doesn't trigger on empty streams.
       We need something like
        .accumulate(Function<ConstraintStream, UniConstraintStream<T>>, T defaultValue): ConstraintStream+1
       To change it from 3 separate constraints (one for the normal case, one in the case of empty stream,
       one to remove double counting).
     */
```

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java:129]: Another static import on `Joiners`?

**Review by triceo — COMMENTED** [review]
One last going through the code.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/BusPassengerCountTotalUpdatingVariableListener.java
Comment (Christopher-Chianelli) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/BusPassengerCountTotalUpdatingVariableListener.java:74]: As a side note, can a bus stop have 0 passengers? I don't think it can (and if it does, it an error; passengerQuantity is NOT a shadow variable or a planning variable)

Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/BusPassengerCountTotalUpdatingVariableListener.java:74]: Good point. Maybe we should remove that check altogether.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java
Comment (Christopher-Chianelli) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/solver/CoachShuttleGatheringConstraintProvider.java:42]: No, all constraint providers in `optaplanner-examples` use `ScoreType.field`, not a static import of field from ScoreType.

**Review by triceo — APPROVED** [review]
LGTM. I'll merge tomorrow afternoon unless there's any more comments by then.
Good job, Chris!

**Comment by triceo** [comment]
Sonar makes a good point - this PR is missing a test for the constraint provider.
Would you mind adding one, @Christopher-Chianelli?

**Comment by triceo** [comment]
Also, the 6 code smells are trivial to address.

**Comment by Christopher-Chianelli** [comment]
@triceo I'll add a test for the constraint provider tomorrow

**Pull request merged**
